### PR TITLE
Add support for PSR-18 HTTP clients

### LINKED ISSUE #520: Migration to a HTTP library
Roadmap declares the intention to move from custom web fetcher to something more standardized. At selfoss, we are currently [in the process](https://github.com/SSilence/selfoss/pull/936) of doing the same so we would like to help with this for our dependencies, including SimplePie.

Since selfoss targets PHP 5.4, we would need to use older version of Guzzle (5.*), which might cause conflicts for applications using SimplePie and a different version of Guzzle. For this reason I propose depending on a [HTTPlug](http://httplug.io/) interface and letting the application choose the network library.

If you do not have any objections I will start working on it.

### PR DESCRIPTION
**Pull request by Art4**
Hi all 👋 

This is a follow up of #774 and will add support for every [PSR-18](https://www.php-fig.org/psr/psr-18/) HTTP client implementation. 🥳 

This PR will fixes #520. I propose this PR for SimplePie 1.9.0, see #731.

### CODE REVIEW

File: src/Content/Type/Sniffer.php [hunk 1]
```diff
@@ -71,8 +73,17 @@ class Sniffer
      *
      * @param File $file Input file
      */
-    public function __construct(File $file)
+    public function __construct(/* File */ $file)
```
Comment (Art4) [inline src/Content/Type/Sniffer.php:76]: Removing the type hint is no breaking change because we have not released 1.9.0

**Comment by Art4** [comment]
This PR is ready for review. Because this PR is build on #774 I recommend to merge this PR first. If will fix some BC breaks that comes with #774.

To show only the changes between #774 and #777 you can look at the diff between the branches: https://github.com/Art4/simplepie/compare/split-file-into-client-and-response...Art4:simplepie:add-psr18-http-client-support

File: src/HTTP/Psr18Client.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+                    break;
+                }
+
+                $remainingRedirects--;
+                $followRedirect = true;
+
+                $requestedUrl = $response->getHeaderLine('Location');
+
+                if ($statusCode === 301) {
+                    $permanentUrl = $requestedUrl;
```
Comment (jtojnar) [inline src/HTTP/Psr18Client.php:120]: We need some kind of mechanism to not change the permanent URL after temporary redirects (as tested by https://github.com/simplepie/simplepie/pull/819). For example, this is what File does:

https://github.com/simplepie/simplepie/blob/e356f5dc711ad4c5d92091a7dea509bf7c1c11a6/src/File.php#L144-L144

**Review by jtojnar — CHANGES_REQUESTED** [review]
Still need to resolve the question of [non-absolute Location header](https://github.com/simplepie/simplepie/pull/826#discussion_r1142088339) and [Redirect tracking](https://github.com/simplepie/simplepie/pull/826#discussion_r1227293061).

**Comment by mblaney** [comment]
thanks @Art4 this looks good, can we merge and handle the redirect case separately?

**Comment by Art4** [comment]
@mblaney I resolved all conflicts and fixed all PHPStan errors.

**Pull request merged**